keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,594 @@
+larger than X.
+- The kubernetes/release tooling must build kubeadm version X from the
+kubeadm repository while build older kubeadm from k/k.
+
+#### Releasing and tagging
+
+##### Challenges
+
+The current plan is for kubeadm to continue to follow the k8s version scheme
+and release schedule. This will continue to make the skew of kubeadm
+and k8s easy to understand for the users and continue to give motivation
+to the kubeadm developers to get work done on time.
+
+This creates some complications:
+- Automation tools have to build kubeadm as part of the release
+- The k/kubeadm tags and branches need to follow those of k/k
+- k/k tags are not SEMVER, which can cause troubles when importing
+the k/kubeadm repository
+
+Currently kubeadm is released as part of the following k8s artifacts:
+- Node tarballs (Linux / Windows)
+- deb/rpm packages
+
+The kubeadm binary is built using either bazel (amd64 only) or using Docker
+(cross-build, all supported architectures).
+
+The binary is built with the same version as the rest of the released binaries
+like the kubelet. The version tooling resides in k/k under the /hack directory
+and it must be copied to k/kubeadm for consistency. The tooling that perform
+the packaging and building are location in k/release and require manual interaction.
+
+##### Bazel vs go build
+
+Bazel was investigated as an option for building kubeadm from the new
+repository but the benefits are unclear. BUILD files present a maintenance
+challenge and also the bazel WORKSPACE and gazelle files need to be kept
+in sync with k8s. On the other hand building using “go build” can take
+around ~20 sec on a fast machine with fast internet.
+
+The majority voted to not use bazel in the k/kubeadm repository.
+
+If caching is required a build container can be used.
+
+##### Syncing tags and branches between k/k and k/kubeadm
+
+A brief discussion of the #prow maintainers revealed the proposal that
+a post-submit job can be used for that instead of prow plugins.
+
+Supposedly such a job will trigger each time a new branch is created
+or if a tag is pushed. This means that the post-submit can trigger any bash
+script or application to get the latest list of branches and tags in k/k
+and apply the same to the kubeadm repository.
+
+Given kubeadm will potentially move in the 1.18 cycle this means that not
+all tags and branches have to be synced - see the options in “Versions
+in the support skew”. So potentially there could be a minimum version
+check in the presubmit.
+
+Scenarios:
+
+###### k/k creates a new release-x.yy branch
+
+The post-submit job checks out master in k/kubeadm, creates a new branch
+release-x.yy and pushes the branch.
+
+###### k/k ‘s release-x.yy branch is fast-forwarded to master
+
+Currently this is a manual step performed by someone on the release team.
+It might be something that the k/kubeadm maintainers have to do manually
+on a daily basis. Automation is possible but is tricky and needs further
+discussion.
```

Comment:
What does doing on a daily basis mean here? Like checking on a daily basis if a fast-forward is necessary? I presume this fast-forward doesn't happen very frequently on k/k, right?

Reply:
> Like checking on a daily basis if a fast-forward is necessary?

yes.

>  I presume this fast-forward doesn't happen very frequently on k/k, right?

https://github.com/kubernetes/sig-release/blob/master/release-engineering/role-handbooks/branch-manager.md#branch-fast-forward

> This is done daily as soon as the release-x.y branch has been cut (which happens after beta.0 is released).

Reply:
on paper, yes.
in reality k/k and SIG Release don't want to automate it yet.

the automation resulting in unwanted commits in a release branch is not something we want to have to deal with.

Reply:
let's keep this for followup iterations on the KEP - potential automation improvements for the future.